snakesist/exist_client.py

```diff
@@ -195,22 +172,117 @@ def __init__(
+
+        user = parsed_url.username or ""
+        password = parsed_url.password or ""
+        host = parsed_url.hostname
+        port = parsed_url.port
+
+        if not isinstance(host, str) and host:
+            raise ValueError(f"Invalid host in URL: {host}")
+
+        if transport is None:
+            probe_result = cls._probe_transport_and_port(
+                f"{user}:{password}@{host}", port
+            )
+            if probe_result is None:
+                raise RuntimeError(f"Couldn't figure out how to talk to {host}.")
+            transport, port = probe_result
+        elif port is None:
+            port = TRANSPORT_PROTOCOLS[transport]
+
+        prefix = cls._probe_instance_prefix(
+            f"{transport}://{user}:{password}@{host}:{port}", parsed_url.path
+        )
+        if prefix is None:
+            raise RuntimeError(
+                "Couldn't determine the location of the 'rest' interface."
+            )
+
+        return cls(
+            transport=transport,
+            host=host,
+            port=port,
+            user=user,
+            password=password,
+            prefix=prefix,
+            parser=parser,
+        )
 
     @staticmethod
-    def _join_paths(*args):
-        return "/".join(s.strip("/") for s in args)
+    def _probe_transport_and_port(
+        host: str, port: Optional[int]
+    ) -> Optional[Tuple[str, int]]:
+        for transport, default_port in TRANSPORT_PROTOCOLS.items():
+            _port = port or default_port
+            try:
+                requests.head(f"{transport}://{host}:{_port}/")
+            except requests.exceptions.ConnectionError:
+                pass
+            else:
+                return transport, _port
+        return None
 
+    @staticmethod
+    def _probe_instance_prefix(base: str, path: str) -> Optional[str]:
+        path_parts = PurePosixPath(path).parts[1:]
+        encountered_collection = False
+
+        # looks for longest path as different instances could have overlapping prefixes
+        # will return false results if a path contained a part named "rest"
+        for i in range(len(path_parts), 0, -1):
+            response = requests.get(f"{base}/{'/'.join(path_parts[:i])}/rest/")
+
+            if not content_type_is_xml(response.headers.get("Content-Type", "")):
+                if encountered_collection:
+                    break
+                else:
+                    continue
+
+            qualified_name = etree.QName(etree.fromstring(response.content))
+            if (
+                qualified_name.namespace == EXISTDB_NAMESPACE
+                and qualified_name.localname == "result"
+            ):
+                encountered_collection = True
+            elif encountered_collection:
+                break
+
+        if encountered_collection:
+            return "/".join(path_parts[:i])
+        else:
+            return None
+
+    # TODO there's only one usage
```

Comment:
not very sure what to do here. would refactoring the method away be a good idea? i actually don't see why not and i can't think rn of any use case in the very near future. if there's ever a need for this method, it can be put back in...

same also applies for the other todo on line 309.

Reply:
yep, i annotated these, b/c i wasn't sure about your architectural sentiment. at this point they just eat stack space and produce additional instructions while either the called or the calling methods aren't too complex.

Reply:
yep, i agree. i'll move their contents to their respective calls.